### PR TITLE
Adding google_compute_service_attachment data source

### PR DESCRIPTION
**Pull request by sanmaym**
This PR is for a new data resource that obtains the URI for a service attachment. Its based of this [google api ](https://cloud.google.com/compute/docs/reference/rest/v1/serviceAttachments/get)

[Issue is tracked here] (https://github.com/hashicorp/terraform-provider-google/issues/17122)

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-datasource
`google_compute_service_attachment`
```

### CODE REVIEW

File: mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+	"github.com/hashicorp/terraform-provider-google/google/acctest"
+	"github.com/hashicorp/terraform-provider-google/google/tpgresource"
+)
+
+func TestAccDataSourceGoogleServiceAttachment(t *testing.T) {
+	t.Parallel()
```
Comment (NickElliot) [inline mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go:16]:
```suggestion
	t.Parallel()
	
	context := map[string]interface{}{
		"random_suffix": acctest.RandString(t, 10),
	}
```

File: mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+	serviceAttachmentName := fmt.Sprintf("tf-test-%s", acctest.RandString(t, 10))
```
Comment (NickElliot) [inline mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go:18]:
```suggestion
```

File: mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+	acctest.VcrTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.AccTestPreCheck(t) },
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories(t),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccDataSourceGoogleServiceAttachmentConfig(serviceAttachmentName),
```
Comment (NickElliot) [inline mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go:24]:
```suggestion
				Config: testAccDataSourceGoogleServiceAttachmentConfig(context),
```

File: mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+		if !tpgresource.CompareSelfLinkOrResourceName("", ds_attr["self_link"], rs_attr["self_link"], nil) && ds_attr["self_link"] != rs_attr["self_link"] {
+			return fmt.Errorf("self link does not match: %s vs %s", ds_attr["self_link"], rs_attr["self_link"])
+		}
+
+		return nil
+	}
+}
+
+func testAccDataSourceGoogleServiceAttachmentConfig(name string) string {
```
Comment (NickElliot) [inline mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go:72]:
```suggestion
func testAccDataSourceGoogleServiceAttachmentConfig(context map[string]interface{}) string {
```

File: mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+  resource "google_compute_region_backend_service" "producer_service_backend" {
+	name   = "producer-service"
+	region = "us-west1"
+  
+	health_checks = [google_compute_health_check.producer_service_health_check.id]
+  }
+  
+  resource "google_compute_health_check" "producer_service_health_check" {
+	name = "producer-service-health-check"
```
Comment (NickElliot) [inline mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go:103]:
```suggestion
	name = "producer-service-health-check%{random_suffix}"
```

File: mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+  
+	network       = google_compute_network.psc_ilb_network.id
+	purpose       =  "PRIVATE_SERVICE_CONNECT"
+	ip_cidr_range = "10.1.0.0/16"
+  }
+data "google_compute_service_attachment" "my_attachment" {
+  name = google_compute_service_attachment.foobar.name
+  region = google_compute_service_attachment.foobar.region
+}
+`, name)
```
Comment (NickElliot) [inline mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go:136]:
```suggestion
`, context)
```

File: mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+		if !tpgresource.CompareSelfLinkOrResourceName("", ds_attr["self_link"], rs_attr["self_link"], nil) && ds_attr["self_link"] != rs_attr["self_link"] {
+			return fmt.Errorf("self link does not match: %s vs %s", ds_attr["self_link"], rs_attr["self_link"])
+		}
+
+		return nil
+	}
+}
+
+func testAccDataSourceGoogleServiceAttachmentConfig(name string) string {
+	return fmt.Sprintf(`
```
Comment (NickElliot) [inline mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go:73]:
```suggestion
	return acctest.Nprintf(`
```

File: mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+  description              = "my-description"
+  region                   = "us-west1"
+  enable_proxy_protocol    = true
+  connection_preference    = "ACCEPT_AUTOMATIC"
+  nat_subnets              = [google_compute_subnetwork.psc_ilb_nat.id]
+  target_service           = google_compute_forwarding_rule.psc_ilb_target_service.id
+}
+
+resource "google_compute_forwarding_rule" "psc_ilb_target_service" {
+	name   = "producer-forwarding-rule"
```
Comment (NickElliot) [inline mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go:85]:
```suggestion
	name   = "producer-forwarding-rule%{random_suffix}"
```

File: mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+		}
+
+		return nil
+	}
+}
+
+func testAccDataSourceGoogleServiceAttachmentConfig(name string) string {
+	return fmt.Sprintf(`
+resource "google_compute_service_attachment" "foobar" {
+  name                     = "%s"
```
Comment (NickElliot) [inline mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go:75]:
```suggestion
  name                     = "tf-test%{random_suffix}"
```

File: mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+	name = "producer-service-health-check"
+  
+	check_interval_sec = 1
+	timeout_sec        = 1
+	tcp_health_check {
+	  port = "80"
+	}
+  }
+  resource "google_compute_network" "psc_ilb_network" {
+	name = "psc-ilb-network"
```
Comment (NickElliot) [inline mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go:112]:
```suggestion
	name = "psc-ilb-network%{random_suffix}"
```

File: mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+	auto_create_subnetworks = false
+  }
+  
+  resource "google_compute_subnetwork" "psc_ilb_producer_subnetwork" {
+	name   = "psc-ilb-producer-subnetwork"
```
Comment (NickElliot) [inline mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go:117]:
```suggestion
	name   = "psc-ilb-producer-subnetwork%{random_suffix}"
```

File: mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+	region = "us-west1"
+  
+	network       = google_compute_network.psc_ilb_network.id
+	ip_cidr_range = "10.0.0.0/16"
+  }
+  
+  resource "google_compute_subnetwork" "psc_ilb_nat" {
+	name   = "psc-ilb-nat"
```
Comment (NickElliot) [inline mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go:125]:
```suggestion
	name   = "psc-ilb-nat%{random_suffix}"
```

**Review by NickElliot — CHANGES_REQUESTED** [review]
Hi, the test is having an issue due to conflicting resource names. I suggested a broad formatting adjustment that should resolve this. Thanks!

**Comment by sanmaym** [comment]
Yes, I understood what i should do next. I was just busy with some other
work. I will make these changes
thank you so much.

*Sanmay Mishra*

Infrastructure Cloud Consultant

804-319-9657 <(804)%319-9657> Mobile




*“I choose to work flexibly & send emails outside normal office hours. **Please
do not feel obligated to respond out of your normal working hours."*

**Comment by sanmaym** [comment]
I accepted all your changes. Is there anything else remaining? It says that the community-checker step is queued

**Review by sanmaym — COMMENTED** [review]
accepted all suggestions

**Comment by sanmaym** [comment]
@NickElliot can I get some guidance on this? it passed locally, but not sure where its failing. 
Bwdisw363uNL2Pj

Also, every time I build the provider, I used to get this error message and always with these 2 resources/files . I would proceed to just replacing the file with the latest source code in github and then it would proceed with running the test case command. 

KRmqkykYviok3Rh

**Comment by NickElliot** [comment]
@sanmaym there is almost certainly some conflict in your setup with your terraform/rebase -- often this sort of issue is resolved by doing the following:
clean your tpg/b folders, git pull
rebase your mmv1 branch on main
build mmv1

now you should be able to build the provider. if your tpg version gets out of date with your mmv1 version, it can end up causing funky build problems locally.

**Review by NickElliot — APPROVED** [review]
As is all looks good though! Just run the gofmt on the files and this PR will be ready to go.

**Comment by sanmaym** [comment]
> @sanmaym there is almost certainly some conflict in your setup with your terraform versions -- often this sort of issue is resolved by doing the following: clean your tpg/b folders, git pull rebase your mmv1 branch on main build mmv1
> 
> now you should be able to build the provider. if your tpg version gets out of date with your mmv1 version, it can end up causing funky build problems locally even when nothing is wrong with the code (as seen by the CI passing your test)

@NickElliot I'm facing some challenges now. When I pull from upstream to get latest tpg updates, it didnt work. So, I just manually copied over the tpg directory from remote main to my local main. When I tried to build again, i'm seeing this error message

```
 for terraform
I, [2024-02-12T00:34:04.218685 #18239]  INFO -- : Compiling common files for terraform
make[1]: Leaving directory '/magic-modules'
make tpgtools
make[1]: Entering directory '/magic-modules'
make serialize
make[2]: Entering directory '/magic-modules'
cd tpgtools;\
        cp -f serialization.go.base serialization.go &&\
        go run . --path "api" --overrides "overrides" --mode "serialization" > temp.serial &&\
        mv -f temp.serial serialization.go
**go: cannot find main module, but found .git/config in /magic-modules**
        to create a module there, run:
        cd .. && go mod init
```
Any thoughts? 

Also , I have a second question about your comment here `Just run the gofmt on /google/services/compute/data_source_compute_service_attachment_test.go and this PR will be ready to go.`

This file is in the terraform provider that gets built right? As per my understanding, I only raise the PR from my magic modules repo. Then, what is the purpose of running the gofmt command there?

**Comment by sanmaym** [comment]
any thoughts @NickElliot ?

**Comment by NickElliot** [comment]
sorry for the delay, I went on PTO a couple days before your updates and didn't flag this PR to another reviewer, my bad!

**Review by NickElliot — COMMENTED** [review]
> build issue

Please do the following within your `/github.com/hashicorp/terraform-provider-google-beta/` directory
```
git checkout -- .
git clean -f google/ google-beta/ website/
git pull
```
this should allow you to clean up your provider and get the updates

then in your mmv1 folder while in your fork and branch for this pull request
```
git fetch upstream
git rebase upstream/main
```
then finish the rebase process

>gofmt

`mmv1/third_party/terraform/services/compute/data_source_compute_service_attachment_test.go` is a file you added in this pr, I'm asking you to run it on that file, not the downstream generated one.

**Comment by sanmaym** [comment]
@NickElliot thanks for the detailed review.. does this look good now?

**Comment by NickElliot** [comment]
please rebase again, sorry!

**Review by NickElliot — CHANGES_REQUESTED** [review]
see above

**Pull request closed without merging**